new(config): add `falco_libs.thread_table_size`

incertum opened the pull request:
<!--  Thanks for sending a pull request! Here are some tips for you:
1. If this is your first time, please read our contributor guidelines in the https://github.com/falcosecurity/.github/blob/main/CONTRIBUTING.md file.
2. Please label this pull request according to what type of issue you are addressing.
3. Please add a release note!
4. If the PR is unfinished while opening it specify a wip in the title before the actual title, for example, "wip: my awesome feature"
-->

**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

> /kind bug

> /kind cleanup

> /kind design

> /kind documentation

> /kind failing-test

/kind feature

> /kind release

<!--
Please remove the leading whitespace before the `/kind <>` you uncommented.
-->

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

> /area build

/area engine

> /area tests

> /area proposals

> /area CI

<!--
Please remove the leading whitespace before the `/area <>` you uncommented.
-->

**What this PR does / why we need it**:

We recently increased the default threadtable size to accommodate modern infrastructures better as we have observed drops related to a full internal thread table. This PR makes this setting configurable by the end user.

**Which issue(s) this PR fixes**:

<!--
Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
If PR is `kind/failing-tests` or `kind/flaky-test`, please post the related issues/tests in a comment and do not use `Fixes`.
-->

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

<!--
If NO, just write "NONE" in the release-note block below.

If YES, a release note is required, enter your release note in the block below. 
The convention is the same as for commit messages: https://github.com/falcosecurity/.github/blob/main/CONTRIBUTING.md#commit-convention
If the PR introduces non-backward compatible changes, please add a line starting with "BREAKING CHANGE:" and describe what changed.
For example, `BREAKING CHANGE: the API interface of the rule engine has changed`.
Your note will be included in the changelog.
-->

```release-note
new(config): add `falco_libs.thread_table_size`
```

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/falco/pull/3071#" title="Author self-approved">incertum</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Ffalco).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/falco/blob/master/OWNERS)~~ [incertum]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

incertum commented:
/milestone 0.38.0

incertum commented on falco.yaml:
```diff
@@ -1262,6 +1263,23 @@ base_syscalls:
+# and the fields called `n_drops_full_threadtable` or `n_store_evts_drops` will inform you
+# if you should increase this value for optimal performance.
+state_engine:
+  thread_table_size: 262144
```
@FedeDP and @jasondellaluce since you reviewed the libs changes. Thank you!

FedeDP commented on userspace/falco/configuration.cpp:
```diff
@@ -63,6 +63,7 @@ falco_configuration::falco_configuration():
 	m_syscall_evt_drop_max_burst(1),
 	m_syscall_evt_simulate_drops(false),
 	m_syscall_evt_timeout_max_consecutives(1000),
+	m_state_engine_thread_table_size(262144),
```
Can we have a macro for the default value? Given it is a "weird" number used twice :)

incertum replied:
Agreed, funny enough if you search that number in the Linux kernel it is not so odd.

Before I re-push, was wondering about the Windows build failure. is there something in libs that needs to be adjusted?

FedeDP replied:
> was wondering about the Windows build failure. is there something in libs that needs to be adjusted?

Don't really know why it is complaining :/ 
```
jsoncpp.lib(json_value.obj) : error LNK2019: unresolved external symbol __imp_modf referenced in function "public: bool __cdecl Json::Value::isInt64(void)const " (?isInt64@Value@Json@@QEBA_NXZ) [D:\a\falco\falco\build\libsinsp\examples\sinsp-example.vcxproj]
jsoncpp.lib(json_writer.obj) : error LNK2019: unresolved external symbol __imp__finite referenced in function "class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > __cdecl Json::`anonymous namespace'::valueToString(double,bool,unsigned int,enum Json::PrecisionType)" (?valueToString@?A0x7fc2ca38@Json@@YA?AV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@N_NIW4PrecisionType@2@@Z) [D:\a\falco\falco\build\libsinsp\examples\sinsp-example.vcxproj]
jsoncpp.lib(json_writer.obj) : error LNK2019: unresolved external symbol __imp__isnan referenced in function "class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > __cdecl Json::`anonymous namespace'::valueToString(double,bool,unsigned int,enum Json::PrecisionType)" (?valueToString@?A0x7fc2ca38@Json@@YA?AV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@N_NIW4PrecisionType@2@@Z) [D:\a\falco\falco\build\libsinsp\examples\sinsp-example.vcxproj]
```

Seems a libs issue on jsoncpp; perhaps triggered by: https://github.com/falcosecurity/libs/commit/a89fd5d510fac5b5f2418f263f8d2c4ec80410ef

cc @geraldcombs can you take a look? :pray:  thanks!!

LucaGuerra replied:
Perhaps you could write it as `0x40000` so it's easier to understand. Not sure if you can do the same in yaml as well.

geraldcombs replied:
> Seems a libs issue on jsoncpp

It might be because libs is linking with the dynamic CRT (MultiThreadedDLL) instead of the static one (MultiThreaded). If that's the case we should be able to fix it here by passing `-DCMAKE_MSVC_RUNTIME_LIBRARY=MultiThreaded` to CMake in the `build-win32-package` job, similar to the `build-libs-others-amd64` job in libs.

Alternatively we could try to fix it in libs, e.g. by trying to override CMAKE_MSVC_RUNTIME_LIBRARY in CMakeLists.txt.

incertum replied:
@LucaGuerra internally yes we can default to 0x40000 in the macro, but don't you think for adopters that is even more weird than a normal number? 🙃

geraldcombs replied:
> @geraldcombs whichever you think is the best approach. I can bump libs here to the new latest after we address it in libs if we choose that path.

I opened a libs PR (1695) to link to the static CRT by default. That seemed to make the most sense, since the project favors static builds.

leogr reviewed (commented):
One thought after a quick look.

I would like to group some top-level config keys instead of continuously adding new ones.

I mean, the top-level config keys are growing exponentially, and this may not be ideal for users approaching the configuration.

incertum commented:
> One thought after a quick look.
> 
> I would like to group some top-level config keys instead of continuously adding new ones.
> 
> I mean, the top-level config keys are growing exponentially, and this may not be ideal for users approaching the configuration.

What would it be? It seems that we don't have a new top level category for `state_engine` related configs that may grow over time?

Andreagit97 commented:
maybe we could add a key called `libsinsp` or `internals` that in the future will collect all the libs configurations... for example, we already have the `libs_logger` config that we could move behind this new config (not now of course) WDYT?

FedeDP commented:
@Andreagit97 that's a great idea actually, even if it is exposing some "internal knowledge" about who owns what.
I'd rather call it just `falco-libs:`.

incertum commented:
> @Andreagit97 that's a great idea actually, even if it is exposing some "internal knowledge" about who owns what. I'd rather call it just `falco-libs:`.

`falco_libs` with underscore instead? Perhaps in the future the `libs_logger` can be moved under `falco_libs` as well. It should be less of a breaking change as likely few folks use that setting.

Additional question: Do we know what's wrong with the Windows build?

Andreagit97 commented:
> falco_libs with underscore instead?

Yep, we need to use the underscore

leogr commented:
In general, I wouldn't say I like the idea of exposing internal names to the users (because internals can change, but the external function remains the same). 

So perhaps something like `libs_internals` (to make it explicitly, these are internal settings and may potentially change very often) 

However, `falco_libs` or just `libs` is still for me if the consensus goes toward this.

geraldcombs commented:
> @geraldcombs using the latest libs commit and also rebased w/ falco master, still having the windows failure. Could you take another look? Thank you!

I'm able to replicate the issue here. If I make the following change here
```
--- a/.github/workflows/reusable_build_packages.yaml
+++ b/.github/workflows/reusable_build_packages.yaml
@@ -233,7 +233,7 @@ jobs:
         run: |
           mkdir build
           cd build
-          cmake -DCMAKE_BUILD_TYPE=Release -DMINIMAL_BUILD=On -DUSE_BUNDLED_DEPS=On -DBUILD_FALCO_UNIT_TESTS=On -DFALCO_VERSION=${{ inputs.version }} ..
+          cmake -DCMAKE_BUILD_TYPE=Release -DMINIMAL_BUILD=On -DUSE_BUNDLED_DEPS=On -DBUILD_FALCO_UNIT_TESTS=On -DCMAKE_MSVC_RUNTIME_LIBRARY=MultiThreaded -DFALCO_VERSION=${{ inputs.version }} ..
```
I can build and run `falco_unit_tests.exe`, although I'm not sure why that's necessary. If I build a standalone version of libs here without setting `CMAKE_MSVC_RUNTIME_LIBRARY` I get a static build.

incertum commented:
Thanks @geraldcombs I'll then rebase, thanks!

incertum commented:
> > @geraldcombs using the latest libs commit and also rebased w/ falco master, still having the windows failure. Could you take another look? Thank you!
> 
> I'm able to replicate the issue here. If I make the following change here
> 
> ```
> --- a/.github/workflows/reusable_build_packages.yaml
> +++ b/.github/workflows/reusable_build_packages.yaml
> @@ -233,7 +233,7 @@ jobs:
>          run: |
>            mkdir build
>            cd build
> -          cmake -DCMAKE_BUILD_TYPE=Release -DMINIMAL_BUILD=On -DUSE_BUNDLED_DEPS=On -DBUILD_FALCO_UNIT_TESTS=On -DFALCO_VERSION=${{ inputs.version }} ..
> +          cmake -DCMAKE_BUILD_TYPE=Release -DMINIMAL_BUILD=On -DUSE_BUNDLED_DEPS=On -DBUILD_FALCO_UNIT_TESTS=On -DCMAKE_MSVC_RUNTIME_LIBRARY=MultiThreaded -DFALCO_VERSION=${{ inputs.version }} ..
> ```
> 
> I can build and run `falco_unit_tests.exe`, although I'm not sure why that's necessary. If I build a standalone version of libs here without setting `CMAKE_MSVC_RUNTIME_LIBRARY` I get a static build.

@geraldcombs just double-checking if I interpret this correctly. You or someone else will open a PR and I shall rebase afterwards? Thanks in advance for clarifying!

incertum commented:
@jasondellaluce already bumped libs before hence I could remove it here. However, the CI still seems to be in a not good state (unrelated to these changes).

jasondellaluce commented on userspace/engine/falco_common.h:
```diff
@@ -28,6 +28,8 @@ limitations under the License.
 //
 #define DEFAULT_OUTPUTS_QUEUE_CAPACITY_UNBOUNDED_MAX_LONG_VALUE std::ptrdiff_t(~size_t(0) / 2)
 
+#define DEFAULT_FALCO_LIBS_INTERNALS_THREAD_TABLE_SIZE 0x4000
```
Shouldn't this be `0x40000`? Seems like this is a quite smaller number

jasondellaluce reviewed (commented):
@incertum I have no major concerns here if not the naming decision, as pointed out above. I'm pretty much onboard with the suggestions of the other reviewers.

incertum commented:
> @incertum I have no major concerns here if not the naming decision, as pointed out above. I'm pretty much onboard with the suggestions of the other reviewers.

@jasondellaluce Could you elaborate? I tried to pick a name that addresses all feedback. If this should not be it, what should the final name be? No strong opinion.

leogr commented:
> > > @incertum I have no major concerns here if not the naming decision, as pointed out above. I'm pretty much onboard with the suggestions of the other reviewers.
> > 
> > 
> > @jasondellaluce Could you elaborate? I tried to pick a name that addresses all feedback. If this should not be it, what should the final name be? No strong opinion.
> 
> I think I'd go with either `falco_libs` or `libs_internals`. But again, just my personal taste!

After thinking about this again, I believe `falco_libs` can be nice, especially if we start distributing Falco's libs at some point in the future.

incertum commented:
SGTM, just changed it. Thanks.

FedeDP reviewed (approved):
/approve

poiana commented:
LGTM label has been added.  <details>Git tree hash: 2a6cb1f153d2edb03f88bba04b0f167decafb967</details>

Andreagit97 commented on userspace/falco/configuration.h:
```diff
@@ -136,6 +136,8 @@ class falco_configuration
 
 	uint32_t m_syscall_evt_timeout_max_consecutives;
 
+	std::uint32_t m_falco_libs_thread_table_size;
```
```suggestion
	uint32_t m_falco_libs_thread_table_size;
```

Andreagit97 reviewed (commented):
a minor nit and then we can go! thank you!

poiana commented:
New changes are detected. LGTM label has been removed.

poiana commented:
LGTM label has been added.  <details>Git tree hash: 8be4088d216f0f73aa595b0db4a073ca47c744dd</details>

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/falco/pull/3071#pullrequestreview-1910527787" title="Approved">FedeDP</a>*, *<a href="https://github.com/falcosecurity/falco/pull/3071#" title="Author self-approved">incertum</a>*, *<a href="https://github.com/falcosecurity/falco/pull/3071#pullrequestreview-1916376911" title="Approved">leogr</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Ffalco).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/falco/blob/master/OWNERS)~~ [FedeDP,incertum,leogr]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

Andreagit97 reviewed (approved):
/approve

Andreagit97 commented:
/unhold
